In [1]:
# Importa a biblioteca pandas com o apelido de pd
import pandas as pd

In [2]:
# Caso não tenha o pandas instalado na máquina, execute esta célula para instalar
!pip install pandas

# Geralmente o numpy já vem instalado, mas caso não esteja instalado,
# basta realizar o mesmo comando com o numpy no lugar do pandas

# Relembrando
Vizualizar os dados

In [4]:
# Leitura da base de dados

# Repare que chamamos o pandas por meio do pd
# em seguida utilizamos a função read_csv para 
# especificar qual o arquivo desejamos que seja lido pelo pandas
df = pd.read_csv('banco_aula_7.csv')

In [5]:
# Exibir as 5 primeiras e 5 últimas linhas
df


idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
0       59.0       admin.          NaN  secondary            2343         yes   
1       56.0       admin.          NaN  secondary              45          no   
2       41.0   technician          NaN  secondary            1270         yes   
3       55.0     services          NaN  secondary            2476         yes   
4       54.0       admin.          NaN   tertiary             184          no   
...      ...          ...          ...        ...             ...         ...   
11157   33.0  blue-collar          NaN    primary               1         yes   
11158   39.0     services          NaN  secondary             733          no   
11159   32.0   technician          NaN  secondary              29          no   
11160   43.0   technician          NaN  secondary               0          no   
11161   34.0   technician          NaN  secondary               0          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
0             no    5  may   1042.0          1.0      yes  
1             no    5  may   1467.0          1.0      yes  
2             no    5  may   1389.0          1.0      yes  
3             no    5  may    579.0          1.0      yes  
4             no    5  may    673.0          2.0      yes  
...          ...  ...  ...      ...          ...      ...  
11157         no   20  apr    257.0          1.0       no  
11158         no   16  jun      NaN          4.0       no  
11159         no   19  aug    156.0          2.0       no  
11160        yes    8  may      9.0          2.0       no  
11161         no    9  jul    628.0          1.0       no  

[11162 rows x 12 columns]

In [6]:
# Exibir as primeiras 5 linhas
df.head()

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0      admin.          NaN  secondary            2343         yes   
1   56.0      admin.          NaN  secondary              45          no   
2   41.0  technician          NaN  secondary            1270         yes   
3   55.0    services          NaN  secondary            2476         yes   
4   54.0      admin.          NaN   tertiary             184          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no    5  may   1042.0          1.0      yes  
1         no    5  may   1467.0          1.0      yes  
2         no    5  may   1389.0          1.0      yes  
3         no    5  may    579.0          1.0      yes  
4         no    5  may    673.0          2.0      yes

In [7]:
# Exibir as primeiras 2 linhas
df.head(2)

idade emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0  admin.          NaN  secondary            2343         yes   
1   56.0  admin.          NaN  secondary              45          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no    5  may   1042.0          1.0      yes  
1         no    5  may   1467.0          1.0      yes

In [7]:
# Exibir as últimas 5 linhas
df.tail()

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
11157   33.0  blue-collar          NaN    primary               1         yes   
11158   39.0     services          NaN  secondary             733          no   
11159   32.0   technician          NaN  secondary              29          no   
11160   43.0   technician          NaN  secondary               0          no   
11161   34.0   technician          NaN  secondary               0          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
11157         no   20  apr    257.0          1.0       no  
11158         no   16  jun      NaN          4.0       no  
11159         no   19  aug    156.0          2.0       no  
11160        yes    8  may      9.0          2.0       no  
11161         no    9  jul    628.0          1.0       no

In [8]:
# Exibir as 2 últimas Linhas
df.tail(2)

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
11160   43.0  technician          NaN  secondary               0          no   
11161   34.0  technician          NaN  secondary               0          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
11160        yes    8  may      9.0          2.0       no  
11161         no    9  jul    628.0          1.0       no

### Primeiras Informações da base de dados

In [9]:
# Exibe informações das colunas numéricas
df.describe()

idade  saldo_bancario           dia       duracao   qtd_contato
count  11143.000000    11162.000000  11162.000000  10111.000000  11034.000000
mean      41.150767     1528.538524     15.658036    400.738799      2.444716
std       11.759461     3225.413326      8.420740    352.470409      2.672403
min       18.000000    -6847.000000      1.000000      2.000000      1.000000
25%       32.000000      122.000000      8.000000    166.000000      1.000000
50%       39.000000      550.000000     15.000000    283.000000      2.000000
75%       49.000000     1708.000000     22.000000    535.000000      3.000000
max       85.000000    81204.000000     31.000000   3881.000000     63.000000

In [10]:
# Exibe uma descrição das colunas do tipo object
df.describe(include = "O")
# A primeira descrição 'count' é a quantidade de valores não nulos
# A segunda descrição 'unique' informa quantos valores únicos existem nas colunas
# A terceira descrição 'top' informa qual o valor que mais aparece na coluna
# A quarta descrição 'freq' é em relação à frequência (quantidade de vezes) que o valor top aparece na coluna

emprego estado_civil   educacao possui_casa emprestimo    mes  \
count        11162         1293      11162       11162      11162  11162   
unique          12            1          4           2          2     12   
top     management     divorced  secondary          no         no    may   
freq          2566         1293       5476        5881       9702   2824   

       deposito  
count     11162  
unique        2  
top          no  
freq       5873

In [11]:
df.info()
# Repare que existem algumas linhas com valores nulos

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11162 entries, 0 to 11161
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   idade           11143 non-null  float64
 1   emprego         11162 non-null  object 
 2   estado_civil    1293 non-null   object 
 3   educacao        11162 non-null  object 
 4   saldo_bancario  11162 non-null  int64  
 5   possui_casa     11162 non-null  object 
 6   emprestimo      11162 non-null  object 
 7   dia             11162 non-null  int64  
 8   mes             11162 non-null  object 
 9   duracao         10111 non-null  float64
 10  qtd_contato     11034 non-null  float64
 11  deposito        11162 non-null  object 
dtypes: float64(3), int64(2), object(7)
memory usage: 1.0+ MB


In [8]:
# Quantidade de valores nulos por coluna
df.isnull().sum() 

idade               19
emprego              0
estado_civil      9869
educacao             0
saldo_bancario       0
possui_casa          0
emprestimo           0
dia                  0
mes                  0
duracao           1051
qtd_contato        128
deposito             0
dtype: int64

Verificar quantos valores nulos cada coluna possui é interessante, mas talvez verificar a porcentagem de valores nulos por coluna seja melhor para entender os valores nulos

In [9]:
# Porcentagem de valores nulos por coluna
# Abaixo é realizada uma regra de 3 onde 
# a quantidade de valores nulos é dividida pela
# quantidade de linhas do dataframe e multiplicado por 100
# Resultando na porcentagem de valores nulos por coluna
# Após a regra de 3, é utilizado a função round para exibir
# 4 casas decimais
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.1702
emprego            0.0000
estado_civil      88.4161
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            9.4159
qtd_contato        1.1467
deposito           0.0000
dtype: float64

In [15]:
df.isnull().sum() / len(df) * 100

idade              0.170220
emprego            0.000000
estado_civil      88.416054
educacao           0.000000
saldo_bancario     0.000000
possui_casa        0.000000
emprestimo         0.000000
dia                0.000000
mes                0.000000
duracao            9.415875
qtd_contato        1.146748
deposito           0.000000
dtype: float64

## Valores únicos

Os valores únicos fazem referência à quantidade de valores que apresentam em uma mesma coluna e existem algumas funções para avaliar estes valores 

In [10]:
# Para verificar a quantidade de valores únicos na coluna emprego
df.emprego.nunique()

12

In [11]:
# Para verificar a os valores únicos na coluna emprego
df.estado_civil.unique()
# No python, quando um valor é nulo é representado por 
# NaN (Abreviação de Not a Number)

array([nan, 'divorced'], dtype=object)

In [12]:
# Para verificar a os valores únicos na coluna emprego e a quantidade de aparições de cada valor
df.emprego.value_counts()

management       2566
blue-collar      1944
technician       1823
admin.           1334
services          923
retired           778
self-employed     405
student           360
unemployed        357
entrepreneur      328
housemaid         274
unknown            70
Name: emprego, dtype: int64

Importante ressaltar que os 3 métodos acima só podem ser utilizados com apenas uma coluna

In [13]:
# Verificar o tamanho do dataframe (Tabela)
df.shape

(11162, 12)

In [14]:
# quantidade de colunas
df.shape[1]

12

In [22]:
# quantidade de linhas
df.shape[0]

11162

In [23]:
# Existe outro método de verificar a quantidade de linhas do dataframe
# é o mesmo método para verificar o tamanho de uma lista
len(df)

11162

In [24]:
# Listar as colunas do dataframe
df.columns

Index(['idade', 'emprego', 'estado_civil', 'educacao', 'saldo_bancario',
       'possui_casa', 'emprestimo', 'dia', 'mes', 'duracao', 'qtd_contato',
       'deposito'],
      dtype='object')

# Selecionar Colunas

In [25]:
# Para selecionar apenas uma coluna
df.idade

0        59.0
1        56.0
2        41.0
3        55.0
4        54.0
         ... 
11157    33.0
11158    39.0
11159    32.0
11160    43.0
11161    34.0
Name: idade, Length: 11162, dtype: float64

In [26]:
# para selecionar uma ou mais colunas
df[['idade','emprego','estado_civil']]

idade      emprego estado_civil
0       59.0       admin.          NaN
1       56.0       admin.          NaN
2       41.0   technician          NaN
3       55.0     services          NaN
4       54.0       admin.          NaN
...      ...          ...          ...
11157   33.0  blue-collar          NaN
11158   39.0     services          NaN
11159   32.0   technician          NaN
11160   43.0   technician          NaN
11161   34.0   technician          NaN

[11162 rows x 3 columns]

# Index
Cada linha da tabela possuí um index único (identificador único), ou seja, não existem index repetidos

O index não chega a ser uma coluna, mas se comporta como uma, sendo ele a primeira "coluna à esquerda da tabela"

Com o index nós podemos selecionar linhas específicas, ou conjunto de linha, semelhante ao que aprendemos em listas

In [27]:
# Exibe informações a respeito do index da tabela
# Semelhante ao que aprendemos no loop for, temos o valor de início, fim e o passo do index
df.index

RangeIndex(start=0, stop=11162, step=1)

In [28]:
# Para exibir o index podemos utilizar a função que o transforma em lista
list(df.index)

[0,
 1,
 2,
 3,
 4,
 5,
 6,
 7,
 8,
 9,
 10,
 11,
 12,
 13,
 14,
 15,
 16,
 17,
 18,
 19,
 20,
 21,
 22,
 23,
 24,
 25,
 26,
 27,
 28,
 29,
 30,
 31,
 32,
 33,
 34,
 35,
 36,
 37,
 38,
 39,
 40,
 41,
 42,
 43,
 44,
 45,
 46,
 47,
 48,
 49,
 50,
 51,
 52,
 53,
 54,
 55,
 56,
 57,
 58,
 59,
 60,
 61,
 62,
 63,
 64,
 65,
 66,
 67,
 68,
 69,
 70,
 71,
 72,
 73,
 74,
 75,
 76,
 77,
 78,
 79,
 80,
 81,
 82,
 83,
 84,
 85,
 86,
 87,
 88,
 89,
 90,
 91,
 92,
 93,
 94,
 95,
 96,
 97,
 98,
 99,
 100,
 101,
 102,
 103,
 104,
 105,
 106,
 107,
 108,
 109,
 110,
 111,
 112,
 113,
 114,
 115,
 116,
 117,
 118,
 119,
 120,
 121,
 122,
 123,
 124,
 125,
 126,
 127,
 128,
 129,
 130,
 131,
 132,
 133,
 134,
 135,
 136,
 137,
 138,
 139,
 140,
 141,
 142,
 143,
 144,
 145,
 146,
 147,
 148,
 149,
 150,
 151,
 152,
 153,
 154,
 155,
 156,
 157,
 158,
 159,
 160,
 161,
 162,
 163,
 164,
 165,
 166,
 167,
 168,
 169,
 170,
 171,
 172,
 173,
 174,
 175,
 176,
 177,
 178,
 179,
 180,
 181,
 182,
 183,
 184,


# Função iloc
com esta função, define-se qual index deseja exibir, ou qual faixa de index deseja

In [29]:
# Exibir as 5 primeiras linhas da tabela
df.head()

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0      admin.          NaN  secondary            2343         yes   
1   56.0      admin.          NaN  secondary              45          no   
2   41.0  technician          NaN  secondary            1270         yes   
3   55.0    services          NaN  secondary            2476         yes   
4   54.0      admin.          NaN   tertiary             184          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no    5  may   1042.0          1.0      yes  
1         no    5  may   1467.0          1.0      yes  
2         no    5  may   1389.0          1.0      yes  
3         no    5  may    579.0          1.0      yes  
4         no    5  may    673.0          2.0      yes

In [32]:
# Selecionando apenas um index, as informações serão exibidas em formato de lista

# Seleciona as informações da tabela onde o index é 0
df.iloc[0]

idade                  59.0
emprego              admin.
estado_civil            NaN
educacao          secondary
saldo_bancario         2343
possui_casa             yes
emprestimo               no
dia                       5
mes                     may
duracao              1042.0
qtd_contato             1.0
deposito                yes
Name: 0, dtype: object

In [31]:
# Selecionando apenas um index, as informações serão exibidas em formato de lista

# Seleciona as informações da dabela onde o index é 1
df.iloc[1]

idade                  56.0
emprego              admin.
estado_civil            NaN
educacao          secondary
saldo_bancario           45
possui_casa              no
emprestimo               no
dia                       5
mes                     may
duracao              1467.0
qtd_contato             1.0
deposito                yes
Name: 1, dtype: object

In [33]:
# Para selecionar mais de um index, utiliza-se a notação [index_inicial:index_final]
# O index final não será exibido (semelhante com o que aprendemos em listas)
df.iloc[0:1]

idade emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0  admin.          NaN  secondary            2343         yes   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no    5  may   1042.0          1.0      yes

In [34]:
# Selecionando os valores de 0 até 5, serão exibidos os index de 0 até 4
df.iloc[0:5]

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0      admin.          NaN  secondary            2343         yes   
1   56.0      admin.          NaN  secondary              45          no   
2   41.0  technician          NaN  secondary            1270         yes   
3   55.0    services          NaN  secondary            2476         yes   
4   54.0      admin.          NaN   tertiary             184          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no    5  may   1042.0          1.0      yes  
1         no    5  may   1467.0          1.0      yes  
2         no    5  may   1389.0          1.0      yes  
3         no    5  may    579.0          1.0      yes  
4         no    5  may    673.0          2.0      yes

In [35]:
# Outra forma de selecionar index é passando uma lista de index para a função iloc
df.iloc[[0,2,4]]

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0      admin.          NaN  secondary            2343         yes   
2   41.0  technician          NaN  secondary            1270         yes   
4   54.0      admin.          NaN   tertiary             184          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no    5  may   1042.0          1.0      yes  
2         no    5  may   1389.0          1.0      yes  
4         no    5  may    673.0          2.0      yes

Isto que estamos fazendo é conhecido como slice (fatiar), pois estamos selecionando somente uma fatia dos dados

Em muitos casos não é necessaŕio utilizar toda a base de dados, para isto, podemos criar uma nova variável para receber este slice (fatia)

In [36]:
# Cria a variável fatia que irá receber os valores com índice de 2 à 4
fatia = df.iloc[2:5]

# Exibe a variável fatia
fatia 

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
2   41.0  technician          NaN  secondary            1270         yes   
3   55.0    services          NaN  secondary            2476         yes   
4   54.0      admin.          NaN   tertiary             184          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
2         no    5  may   1389.0          1.0      yes  
3         no    5  may    579.0          1.0      yes  
4         no    5  may    673.0          2.0      yes

In [37]:
# Cria a variável fatia_2 que irá receber os valores com índice de 241 à 246
fatia_2 = df.iloc[241:247]

# Exibe a variável fatia_2
fatia_2 

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
241   37.0   management          NaN   tertiary             240          no   
242   46.0       admin.          NaN  secondary             849         yes   
243   43.0  blue-collar          NaN  secondary             318         yes   
244   27.0   management          NaN   tertiary             703         yes   
245   38.0     services     divorced  secondary              61          no   
246   53.0    housemaid          NaN    primary             104          no   

    emprestimo  dia  mes  duracao  qtd_contato deposito  
241         no    5  jun   1123.0          1.0      yes  
242         no    5  jun    508.0          2.0      yes  
243         no    5  jun    736.0          2.0      yes  
244         no    5  jun    590.0          3.0      yes  
245         no    5  jun    636.0          1.0      yes  
246        yes    5  jun    701.0          1.0      yes

# Selecionar linhas com o loc

Assim como o iloc, nós podemos selecionar linhas com o loc, mas seu funcionamento não se da pelo index. No Loc, nós definimos para selecionar todos os registros que estejam de acordo com um determinado valor, em uma determinada coluna, ou seja, por exemplo, podemos selecionar todos os registros que possuemque possuem idade igual a 37

Para usar o loc, utilizamos a seguinte sintaxe:
* `df.loc[df['nome_da_coluna'] == valor_desejado]` 

In [38]:
# Selecionar todos os registros com iade igual a 37
df.loc[df['idade'] == 38]

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
10      38.0       admin.          NaN  secondary             100         yes   
38      38.0   management          NaN   tertiary             119         yes   
142     38.0       admin.     divorced  secondary             198         yes   
174     38.0   management     divorced   tertiary            4692         yes   
178     38.0   technician          NaN  secondary             226         yes   
...      ...          ...          ...        ...             ...         ...   
11051   38.0   management     divorced   tertiary             441         yes   
11061   38.0  blue-collar          NaN  secondary             530         yes   
11063   38.0   management          NaN   tertiary             590          no   
11092   38.0     services          NaN   tertiary               0         yes   
11143   38.0   unemployed     divorced  secondary             189         yes   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
10            no    7  may    786.0          1.0      yes  
38            no   13  may    568.0          4.0      yes  
142           no   26  may    431.0          2.0      yes  
174           no   29  may   1363.0          2.0      yes  
178           no   30  may    762.0          3.0      yes  
...          ...  ...  ...      ...          ...      ...  
11051         no   30  jan     67.0          1.0       no  
11061         no   20  apr     59.0          3.0       no  
11063         no   31  jul     49.0          2.0       no  
11092         no   17  apr    239.0          1.0       no  
11143        yes   21  jul      NaN          2.0       no  

[353 rows x 12 columns]

Assim como vimos anteriormente, podemos usar a função .shape[0] para receber quantos registros existem para o determinado filtro

In [39]:
# Selecionar a quantidade de registros com iade igual a 37
df.loc[df['idade'] == 37].shape[0]

370

Também podemos usar a função loc com os valores textuais

In [40]:
# Selecionar todos os registros com o estado civil divorciado
df.loc[df['estado_civil'] == 'divorced']

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
7       60.0     retired     divorced  secondary             545         yes   
15      35.0  management     divorced   tertiary            3837         yes   
19      49.0      admin.     divorced  secondary             168         yes   
20      28.0      admin.     divorced  secondary             785         yes   
22      43.0  management     divorced   tertiary             388         yes   
...      ...         ...          ...        ...             ...         ...   
11135   36.0    services     divorced  secondary             115          no   
11140   55.0    services     divorced  secondary            2166         yes   
11142   52.0      admin.     divorced  secondary              48          no   
11143   38.0  unemployed     divorced  secondary             189         yes   
11145   60.0     retired     divorced   tertiary            -134          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
7             no    6  may   1030.0          1.0      yes  
15            no    8  may   1084.0          1.0      yes  
19           yes    8  may    513.0          1.0      yes  
20            no    8  may    442.0          2.0      yes  
22            no    8  may   2087.0          2.0      yes  
...          ...  ...  ...      ...          ...      ...  
11135         no    6  jun    201.0          1.0       no  
11140         no   28  may     23.0         17.0       no  
11142        yes   17  nov    591.0          1.0       no  
11143        yes   21  jul      NaN          2.0       no  
11145         no   12  may    243.0          1.0       no  

[1293 rows x 12 columns]

In [41]:
# Selecionar a quantidade de registros com o estado civil divorciado
df.loc[df['estado_civil'] == 'divorced'].shape[0]

1293

In [42]:
# Para tirar a prova real, usamos a função value_counts 
# para contar a quantidade de vezes que cada
# valor aparece na coluna estado_civil
df.estado_civil.value_counts()

estado_civil
divorced    1293
Name: count, dtype: int64

# Outra funcionalidade da função loc

Com o loc, podemos passar mais alguns parâmetros, permitindo mudar valores nas linhas em que o filtro selecionou. Veja o exemplo para ficar mais claro para `Alterar os valores da coluna estado_civil`

In [43]:
# Primeiro, vamos verificar quais são os valores únicos da coluna emprego
df.emprego.unique()

array(['admin.', 'technician', 'services', 'management', 'retired',
       'blue-collar', 'unemployed', 'entrepreneur', 'housemaid',
       'unknown', 'self-employed', 'student'], dtype=object)

In [44]:
# Seleciona todos os registros com o valor retired em emprego
df.loc[df['emprego'] == 'retired']

idade  emprego estado_civil   educacao  saldo_bancario possui_casa  \
7       60.0  retired     divorced  secondary             545         yes   
52      59.0  retired          NaN  secondary            1521         yes   
57      60.0  retired          NaN    primary            1542         yes   
72      41.0  retired     divorced    primary            -233         yes   
119     34.0  retired          NaN    primary            4499          no   
...      ...      ...          ...        ...             ...         ...   
11110   35.0  retired          NaN    primary             434          no   
11124   42.0  retired          NaN    primary               0         yes   
11137   49.0  retired          NaN    unknown             247          no   
11144   60.0  retired          NaN  secondary               0         yes   
11145   60.0  retired     divorced   tertiary            -134          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
7             no    6  may   1030.0          1.0      yes  
52           yes   14  may   1125.0          2.0      yes  
57            no   14  may    930.0          1.0      yes  
72            no   15  may   1156.0          2.0      yes  
119           no   21  may   1681.0          1.0      yes  
...          ...  ...  ...      ...          ...      ...  
11110         no    5  may    138.0          1.0       no  
11124         no    8  may    186.0          2.0       no  
11137         no   15  may    213.0          2.0       no  
11144         no   15  may    229.0          1.0       no  
11145         no   12  may    243.0          1.0       no  

[778 rows x 12 columns]

In [46]:
# após realizar o filtro selecionando os registros que possuem o valor 'married'
# Colocamos uma vírgula e selecionamos qual coluna desejamos realizar a alteração
# Neste caso selecionamos a própria coluna estado_civil, mas poderia ser qualquer 
# outra coluna.
# Após definir a coluna, fechamos o colchetes e, com o sinal de igual, definimos
# Qual será o novo valor para a coluna estado_civil onde os registros eram 'married'
df.loc[df['emprego'] == 'retired','emprego'] = 'Aposentado'

In [47]:
# Verificando novamente os valores presentes na coluna estado_civil
# Repare que não existe mais o valor married, pois todos os registros 
# Que possuiam este valor foram alterados para 'Casado'
df.emprego.unique()

array(['admin.', 'technician', 'services', 'management', 'Aposentado',
       'blue-collar', 'unemployed', 'entrepreneur', 'housemaid',
       'unknown', 'self-employed', 'student'], dtype=object)

# Valores nulos
Voltando a falar de valores nulos, nas colunas em que existem valores nulos, existem algumas observações a serem feitas:
* Qual o tipo de dados da coluna
* O que a coluna representa
* Qual a porcentagem de dados nulos

In [48]:
# Porcentagem de valores nulos por coluna
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.1702
emprego            0.0000
estado_civil      88.4161
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            9.4159
qtd_contato        1.1467
deposito           0.0000
dtype: float64

Na base de dados, existem 4 colunas com alguns valores nulos.
Destas, as colunas com valores numéricos são:
* idade;
* duracão;
* qtd_contato.

E a coluna estado_civil é do tipo textual

Nas colunas com tipo de dados numéricos, podemos utilizar 3 abordagens:
* 1 - Se forem poucos dados nulos, apagar os registros com estes dados;
* 2 - Se for uma quantidade considerável de dados podemos:
    * 2.1 - Substituir os valores nulos pela média da coluna;
    * 2.1 - Substituir os valores nulos pela mediana da coluna;
    * 2.1 - Substituir os valores nulos pela moda da coluna.
* 3 - Se forem muitos dados nulos, apagar a coluna.

Nas colunas com tipo de dados textual, podemos utilizar 3 abordagens:
* 1 - Se forem poucos dados nulos, apagar os registros com estes dados;
* 2 - Se for uma quantidade considerável de dados podemos substituir os valores nulos pela moda da coluna;
* 3 - Se forem muitos dados nulos, apagar a coluna.

O que é média, mediana e moda:
* Média - é a soma de todos os valores, dividido pela quantidade de valores somados
* Mediana - considerando a lista de valores em ordem crescente, é o valor central da lista
* Moda - é o valor que mais aparece na coluna

Para fazer as alterações nos valores nulos podemos usar a função loc, se for alterar os valores, ou drop se for apagar a coluna.
No caso da função loc, não utilizamos um sinal de igualdade para verificar os valores nulos e sim, usamos o .isnull()


### Coluna numérica abordagem 1 - Apagar as linhas.
Como na coluna idade não temos nem 0,2% de registros nulos, é aceitável que se exclua tais registros (Poderíamos utilizar os outros métodos também)

In [49]:
# Exibe todos os registros com valores nulos na coluna idade
df.loc[df['idade'].isnull()] 

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
1243     NaN  Aposentado     divorced  secondary               1          no   
1582     NaN  Aposentado     divorced    primary            2282          no   
2598     NaN  Aposentado     divorced    primary               0          no   
2793     NaN  Aposentado     divorced    primary            6746          no   
2802     NaN  Aposentado          NaN    unknown             775          no   
2865     NaN  Aposentado          NaN    primary             230          no   
3002     NaN  Aposentado          NaN  secondary             614          no   
3067     NaN  Aposentado          NaN    primary            2190          no   
3098     NaN  Aposentado          NaN    unknown             775          no   
3292     NaN  Aposentado     divorced    primary             712          no   
3381     NaN  Aposentado          NaN    primary            5236          no   
3716     NaN  Aposentado          NaN    primary            1255          no   
3756     NaN  Aposentado          NaN    unknown             775          no   
3822     NaN  Aposentado          NaN    unknown             775          no   
5521     NaN  Aposentado          NaN  secondary             433          no   
8167     NaN  Aposentado          NaN  secondary             433          no   
9430     NaN  Aposentado          NaN   tertiary             553          no   
10562    NaN  Aposentado          NaN    primary             648          no   
10843    NaN  Aposentado     divorced    unknown             157          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
1243          no   13  feb    152.0          3.0      yes  
1582          no   21  apr    207.0         17.0      yes  
2598          no   26  aug    189.0          1.0      yes  
2793          no   21  oct    749.0          1.0      yes  
2802          no   22  oct    313.0          3.0      yes  
2865          no   30  oct    144.0          1.0      yes  
3002          no    9  dec    595.0          1.0      yes  
3067          no   12  jan    512.0          2.0      yes  
3098          no   26  jan    164.0          4.0      yes  
3292          no    3  mar    557.0          1.0      yes  
3381          no    1  apr    558.0          2.0      yes  
3716          no   14  jul    247.0          1.0      yes  
3756          no   22  jul    860.0          2.0      yes  
3822          no    4  aug    476.0          2.0      yes  
5521          no   15  sep    161.0          1.0       no  
8167          no   15  dec    150.0          1.0       no  
9430          no   19  aug   2027.0          5.0       no  
10562         no    3  sep    318.0          1.0       no  
10843         no    7  sep    147.0          1.0       no

In [50]:
# Retornamos uma lista com o index de todos os
# registros onde a idade é nula 
df.loc[df['idade'].isnull()].index

Index([ 1243,  1582,  2598,  2793,  2802,  2865,  3002,  3067,  3098,  3292,
        3381,  3716,  3756,  3822,  5521,  8167,  9430, 10562, 10843],
      dtype='int64')

In [51]:
# Ao executar o código abaixo foram removidos todos os registros 
# Que possuiam a idade nula
df.drop(df.loc[df['idade'].isnull()].index, inplace=True)

In [52]:
# Exibe todos os registros com valores nulos na coluna idade
df.loc[df['idade'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [53]:
# Porcentagem de valores nulos por coluna
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.0000
emprego            0.0000
estado_civil      88.4501
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            9.4319
qtd_contato        1.1487
deposito           0.0000
dtype: float64

### Coluna numérica abordagem 2 - substituir valores nulos por média, mediana ou moda

In [57]:
round(df.duracao.mean())

401

Importante notar que, a média, geralmente será um valor com casas decimais e, a depender da coluna, devemos remover as casas decimais com a função round.

Por exemplo, a coluna duracao representa quantos dias o cliente tem conta no banco, então não podemos ter que o cliente possui conta há 400.64843 dias...

In [55]:
df.duracao.median()

283.0

In [56]:
# Ao exibir a moda, talvez sejam exibidos alguns valores 
# com a mesma quantidade de ocorrência
df.duracao.mode()

0    119.0
1    158.0
2    161.0
Name: duracao, dtype: float64

Analisando o resultado a cima, averiguamos que os valores 119, 158 e 161 aparecem na base de dados a exata mesma quantidade de vezes, mara verificar isto podemos utilizar a função value_counts(), que retorna em ordem decrescente, os valores que mais aparecem

In [58]:
df.duracao.value_counts()
# Aqui verificamos que os 3 valores aparecem 37 vezes.

duracao
119.0     37
158.0     37
161.0     37
150.0     36
144.0     35
          ..
980.0      1
1158.0     1
1035.0     1
1490.0     1
1812.0     1
Name: count, Length: 1388, dtype: int64

Fica a critério do desenvolvedor, caso vá usar a moda, selecionar qual dos 3 valores irá utilizar para substituir os valores nulos

### Coluna numérica abordagem 2.1 - substituir valores nulos por média

In [59]:
# Retorna a média da coluna
df['duracao'].mean()

400.6484344034879

In [60]:
# Retorna a média arredondada sem casas decimais da coluna
round(df['duracao'].mean())

401

In [61]:
# Exibe todos os registros com valores nulos na coluna duracao
df.loc[df['duracao'].isnull()]

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
838     55.0  blue-collar          NaN  secondary             196          no   
897     44.0       admin.     divorced    primary             429         yes   
900     40.0   technician     divorced   tertiary             568         yes   
901     49.0       admin.          NaN  secondary           10250         yes   
933     34.0       admin.          NaN   tertiary             899         yes   
...      ...          ...          ...        ...             ...         ...   
11136   31.0   management          NaN   tertiary               0         yes   
11143   38.0   unemployed     divorced  secondary             189         yes   
11148   32.0  blue-collar          NaN  secondary               1         yes   
11154   52.0   technician          NaN   tertiary             523         yes   
11158   39.0     services          NaN  secondary             733          no   

      emprestimo  dia  mes  duracao  qtd_contato deposito  
838           no   21  aug      NaN          1.0      yes  
897           no   23  oct      NaN          1.0      yes  
900           no   24  oct      NaN          1.0      yes  
901           no   24  oct      NaN          1.0      yes  
933           no   12  nov      NaN          1.0      yes  
...          ...  ...  ...      ...          ...      ...  
11136         no    8  sep      NaN          1.0       no  
11143        yes   21  jul      NaN          2.0       no  
11148         no    8  may      NaN          1.0       no  
11154        yes    8  jul      NaN          1.0       no  
11158         no   16  jun      NaN          4.0       no  

[1051 rows x 12 columns]

In [62]:
# Altera todos os registros com valores núlos da coluna duração para a média arredondada da coluna
df.loc[df['duracao'].isnull(), 'duracao'] = round(df['duracao'].mean())

In [63]:
# Exibe todos os registros com valores nulos na coluna duracao
df.loc[df['duracao'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [64]:
# Porcentagem de valores nulos por coluna
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.0000
emprego            0.0000
estado_civil      88.4501
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            0.0000
qtd_contato        1.1487
deposito           0.0000
dtype: float64

### Coluna numérica abordagem 2.3 - substituir valores nulos pela moda
Analisando a porcentagem de valores núlos por coluna, ainda temos a coluna numérica qtd_contato

In [65]:
# Neste caso só existe uma moda, que é o valor 1
df.qtd_contato.mode()

0    1.0
Name: qtd_contato, dtype: float64

para retornar o valor propriamente dito com a função mode, usamos algum método para selecionar o valor, como, por exemplo, pegar o index do valor (sempre será zero), pegar o maior valor com a função max() ou o menor com a função min() 

In [66]:
df.qtd_contato.mode()[0]

1.0

In [67]:
df.qtd_contato.mode().max()

1.0

In [68]:
df.qtd_contato.mode().min()

1.0

In [71]:
# Exibe todos os registros com valores nulos na coluna qtd_contato
df.loc[df['qtd_contato'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [70]:
# Altera todos os registros com valores núlos da coluna duração para a moda da coluna
df.loc[df['qtd_contato'].isnull(), 'qtd_contato'] = df.qtd_contato.mode()[0]

In [72]:
# Exibe todos os registros com valores nulos na coluna qtd_contato
df.loc[df['qtd_contato'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [73]:
# Porcentagem de valores nulos por coluna
round(df.isnull().sum() / len(df) * 100, 4)

idade              0.0000
emprego            0.0000
estado_civil      88.4501
educacao           0.0000
saldo_bancario     0.0000
possui_casa        0.0000
emprestimo         0.0000
dia                0.0000
mes                0.0000
duracao            0.0000
qtd_contato        0.0000
deposito           0.0000
dtype: float64

### Coluna de valores textuais 
Na coluna estado_civil temos quase 90% de dados nulos, o melhor que podemos fazer é apagar a coluna, mas, se fossem poucos dados núlos, poderíamos excluir somente as linhas com valores nulos nesta coluna e, ou usar a moda para atribuir o valor que mais aparece na coluna aos valores nulos

In [1]:
# Apenas para exibir a moda da coluna textual, 
# mas não iremos utilizá-la
df.estado_civil.mode()

NameError: name 'df' is not defined

In [ ]:
df.estado_civil.mode()[0]

In [74]:
# Desta forma apenas exibimos a tabela sem a coluna estado_civil
# Devido ao parâmetro inplace ser False
df.drop('estado_civil', inplace=False, axis=1)

idade      emprego   educacao  saldo_bancario possui_casa emprestimo  \
0       59.0       admin.  secondary            2343         yes         no   
1       56.0       admin.  secondary              45          no         no   
2       41.0   technician  secondary            1270         yes         no   
3       55.0     services  secondary            2476         yes         no   
4       54.0       admin.   tertiary             184          no         no   
...      ...          ...        ...             ...         ...        ...   
11157   33.0  blue-collar    primary               1         yes         no   
11158   39.0     services  secondary             733          no         no   
11159   32.0   technician  secondary              29          no         no   
11160   43.0   technician  secondary               0          no        yes   
11161   34.0   technician  secondary               0          no         no   

       dia  mes  duracao  qtd_contato deposito  
0        5  may   1042.0          1.0      yes  
1        5  may   1467.0          1.0      yes  
2        5  may   1389.0          1.0      yes  
3        5  may    579.0          1.0      yes  
4        5  may    673.0          2.0      yes  
...    ...  ...      ...          ...      ...  
11157   20  apr    257.0          1.0       no  
11158   16  jun    401.0          4.0       no  
11159   19  aug    156.0          2.0       no  
11160    8  may      9.0          2.0       no  
11161    9  jul    628.0          1.0       no  

[11143 rows x 11 columns]

In [75]:
# Desta forma apagamos definitivamente a coluna estado_civil
df.drop('estado_civil', inplace=True, axis=1)

In [76]:
# Porcentagem de valores nulos por coluna
# Repare que não temos mais a coluna estado_civil
# E também temos todas as colunas sem valores nulos
round(df.isnull().sum() / len(df) * 100, 4)

idade             0.0
emprego           0.0
educacao          0.0
saldo_bancario    0.0
possui_casa       0.0
emprestimo        0.0
dia               0.0
mes               0.0
duracao           0.0
qtd_contato       0.0
deposito          0.0
dtype: float64

# Salvar alterações

Diferente de uma planilha excel que nós abrimos o arquivo, editamos e salvamos o mesmo arquivo, no python toda alteração realizada na tabela não será salva até que utilize o comando para salvar as alterações.

Para o comando que salva as alterações, é necessário definir um nome para o arquivo. Podemos definir o nome igual ao do arquivo original, o que acarretaria em substituir o arquivo original pelo arquivo alterado, então, por padrão, definimos um nome diferente do original e intuitivo.

Outro ponto importante, por padrão, ao salvar a DataFrame alterado, será criada uma coluna com o index dos registros. Para evitar que isto aconteça, é necessário que chame o parâmetro `index e selecione False`.

In [79]:
df.to_csv('banco_alterado_sem_dados_nulos.csv', index = False)

# Exercício

Quem finalizar o exercício e me explicar o que foi feito em cada coluna ganhará ponto extra

1 - importe a biblioteca pandas com apelido pd

2 - Crie uma variável para receber o banco de dados 'banco_exercicio.csv'

3 - Faça uma análise inicial a respeito da base de dados

4 - Analise o tipo de dados das colunas que possuem valores nulos

5 - Realize o tratamento da base de dados para que nenhuma coluna possua valores nulos

6 - Salve a base de dados com o nome 'banco_exercicios_tratado.csv' e sem o index


Obs: utilize os blocos de código abaixo para realizar isto. Caso necessário, pode criar mais blocos de código

In [78]:
#1
import pandas as pd


In [89]:
#2
df_exercicio = pd.read_csv('banco_exercicio.csv')
df_exercicio

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
0       59.0       admin.      married  secondary          2343.0         yes   
1       56.0       admin.      married  secondary            45.0          no   
2       41.0   technician      married  secondary          1270.0         yes   
3       55.0     services      married  secondary          2476.0         yes   
4       54.0       admin.      married   tertiary           184.0          no   
...      ...          ...          ...        ...             ...         ...   
11157   33.0  blue-collar       single    primary             1.0         yes   
11158   39.0     services      married  secondary           733.0          no   
11159   32.0   technician       single  secondary            29.0          no   
11160   43.0   technician      married  secondary             0.0          no   
11161   34.0   technician      married  secondary             0.0          no   

      emprestimo   dia  mes  duracao  qtd_contato deposito  
0             no   5.0  may   1042.0          1.0      yes  
1             no   5.0  may   1467.0          1.0      yes  
2             no   5.0  may   1389.0          1.0      yes  
3             no   5.0  may    579.0          1.0      yes  
4             no   5.0  may    673.0          2.0      yes  
...          ...   ...  ...      ...          ...      ...  
11157         no  20.0  apr    257.0          1.0       no  
11158         no  16.0  jun     83.0          4.0       no  
11159         no  19.0  aug    156.0          2.0       no  
11160        yes   8.0  may      9.0          2.0       no  
11161         no   9.0  jul    628.0          1.0       no  

[11162 rows x 12 columns]

In [107]:
df_exercicio.head()

idade     emprego estado_civil   educacao  saldo_bancario possui_casa  \
0   59.0      admin.      married  secondary          2343.0         yes   
1   56.0      admin.      married  secondary            45.0          no   
2   41.0  technician      married  secondary          1270.0         yes   
3   55.0    services      married  secondary          2476.0         yes   
4   54.0      admin.      married   tertiary           184.0          no   

  emprestimo  dia  mes  duracao  qtd_contato deposito  
0         no  5.0  may   1042.0          1.0      yes  
1         no  5.0  may   1467.0          1.0      yes  
2         no  5.0  may   1389.0          1.0      yes  
3         no  5.0  may    579.0          1.0      yes  
4         no  5.0  may    673.0          2.0      yes

In [90]:
#3
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             6.6744
emprego           3.0819
estado_civil      0.0000
educacao          4.4526
saldo_bancario    0.1165
possui_casa       0.5107
emprestimo        1.7470
dia               1.6037
mes               1.5051
duracao           0.6361
qtd_contato       4.2734
deposito          0.0000
dtype: float64

In [86]:
#3
df_exercicio.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 11162 entries, 0 to 11161
Data columns (total 12 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   idade           10417 non-null  float64
 1   emprego         10818 non-null  object 
 2   estado_civil    11162 non-null  object 
 3   educacao        10665 non-null  object 
 4   saldo_bancario  11149 non-null  float64
 5   possui_casa     11105 non-null  object 
 6   emprestimo      10967 non-null  object 
 7   dia             10983 non-null  float64
 8   mes             10994 non-null  object 
 9   duracao         11091 non-null  float64
 10  qtd_contato     10685 non-null  float64
 11  deposito        11162 non-null  object 
dtypes: float64(5), object(7)
memory usage: 1.0+ MB


In [88]:
#3
df_exercicio.describe()

idade  saldo_bancario           dia       duracao   qtd_contato
count  10417.000000    11149.000000  10983.000000  11091.000000  10685.000000
mean      41.398963     1528.820074     15.721934    372.269047      2.501919
std       11.936950     3227.098921      8.380991    347.951923      2.653685
min       18.000000    -6847.000000      1.000000      2.000000      1.000000
25%       33.000000      122.000000      8.000000    138.000000      1.000000
50%       39.000000      549.000000     15.000000    255.000000      2.000000
75%       49.000000     1709.000000     22.000000    497.000000      3.000000
max       95.000000    81204.000000     31.000000   3881.000000     43.000000

In [91]:
df_exercicio.idade.mean()

41.39896323317654

In [93]:
round(df_exercicio.idade.mean())

41

## Limpesa de nulos idade

In [97]:
df_exercicio.loc[df_exercicio['idade'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [96]:
#Substituindo valores nulos pela media
df_exercicio.loc[df_exercicio['idade'].isnull(),'idade'] = round(df_exercicio.idade.mean())

In [98]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           3.0819
estado_civil      0.0000
educacao          4.4526
saldo_bancario    0.1165
possui_casa       0.5107
emprestimo        1.7470
dia               1.6037
mes               1.5051
duracao           0.6361
qtd_contato       4.2734
deposito          0.0000
dtype: float64

## Limpeza de nulos: emprego

In [99]:
df_exercicio.loc[df_exercicio['emprego'].isnull()]

idade emprego estado_civil   educacao  saldo_bancario possui_casa  \
47      52.0     NaN      married  secondary          3923.0         yes   
66      49.0     NaN      married    primary           341.0         yes   
246     41.0     NaN      married    primary           104.0          no   
248     56.0     NaN     divorced    primary          1238.0          no   
249     41.0     NaN      married  secondary           879.0          no   
...      ...     ...          ...        ...             ...         ...   
11036   54.0     NaN     divorced   tertiary             9.0          no   
11052   54.0     NaN      married    primary           212.0          no   
11060   58.0     NaN      married    primary           435.0         yes   
11091   34.0     NaN      married   tertiary           677.0          no   
11152   34.0     NaN      married  secondary           390.0         yes   

      emprestimo   dia  mes  duracao  qtd_contato deposito  
47            no  14.0  may    942.0          3.0      yes  
66           yes  15.0  may    520.0          2.0      yes  
246          yes   5.0  jun    701.0          1.0      yes  
248           no   5.0  jun   1558.0          1.0      yes  
249           no   5.0  jun   1053.0          2.0      yes  
...          ...   ...  ...      ...          ...      ...  
11036        yes  29.0  jul    136.0          3.0       no  
11052         no  12.0  aug     81.0          4.0       no  
11060         no  28.0  jul    122.0          6.0       no  
11091         no   9.0  feb     37.0          2.0       no  
11152         no  15.0  jul    659.0          3.0       no  

[344 rows x 12 columns]

In [102]:
df_exercicio.emprego.mode()

0    management
Name: emprego, dtype: object

In [103]:
df_exercicio.loc[df_exercicio['emprego'].isnull(),'emprego'] = df_exercicio.emprego.mode()[0]

In [104]:
df_exercicio.loc[df_exercicio['emprego'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [105]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           0.0000
estado_civil      0.0000
educacao          4.4526
saldo_bancario    0.1165
possui_casa       0.5107
emprestimo        1.7470
dia               1.6037
mes               1.5051
duracao           0.6361
qtd_contato       4.2734
deposito          0.0000
dtype: float64

## Limpeza de nulos: Educação 

In [106]:
df_exercicio.loc[df_exercicio['educacao'].isnull()]

idade      emprego estado_civil educacao  saldo_bancario possui_casa  \
44      52.0   management      married      NaN          2240.0         yes   
58      40.0   technician       single      NaN          3652.0         yes   
65      44.0  blue-collar      married      NaN           146.0         yes   
75      36.0     services       single      NaN           281.0         yes   
108     35.0  blue-collar      married      NaN          4822.0         yes   
...      ...          ...          ...      ...             ...         ...   
11017   42.0  blue-collar      married      NaN          1154.0         yes   
11038   36.0  blue-collar      married      NaN          1023.0         yes   
11059   35.0   technician       single      NaN           726.0         yes   
11072   41.0   technician       single      NaN           356.0         yes   
11137   49.0      retired      married      NaN           247.0          no   

      emprestimo   dia  mes  duracao  qtd_contato deposito  
44            no  13.0  may    845.0          1.0      yes  
58            no  14.0  may   1028.0          2.0      yes  
65            no  15.0  may    483.0          1.0      yes  
75            no  16.0  may    515.0          1.0      yes  
108           no  20.0  may    843.0          2.0      yes  
...          ...   ...  ...      ...          ...      ...  
11017         no  16.0  apr    341.0          2.0       no  
11038         no   4.0  feb    184.0          2.0       no  
11059         no   5.0  may    163.0          1.0       no  
11072         no  13.0  may     93.0          2.0       no  
11137         no  15.0  may    213.0          2.0       no  

[497 rows x 12 columns]

In [108]:
df_exercicio.educacao.mode()

0    secondary
Name: educacao, dtype: object

In [109]:
df_exercicio.loc[df_exercicio['educacao'].isnull(),'educacao'] = df_exercicio.educacao.mode()[0]

In [110]:
df_exercicio.loc[df_exercicio['educacao'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [111]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           0.0000
estado_civil      0.0000
educacao          0.0000
saldo_bancario    0.1165
possui_casa       0.5107
emprestimo        1.7470
dia               1.6037
mes               1.5051
duracao           0.6361
qtd_contato       4.2734
deposito          0.0000
dtype: float64

## Limpeza de nulos: saldo_bancario 

In [112]:
df_exercicio.loc[df_exercicio['saldo_bancario'].isnull()]

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
936     32.0  blue-collar       single   tertiary             NaN          no   
1213    33.0  blue-collar       single   tertiary             NaN          no   
1735    33.0  blue-collar       single   tertiary             NaN          no   
1871    24.0     services      married  secondary             NaN         yes   
3458    30.0       admin.       single  secondary             NaN         yes   
5130    26.0   technician      married  secondary             NaN          no   
5319    48.0   management     divorced   tertiary             NaN         yes   
5857    45.0  blue-collar      married    primary             NaN         yes   
6080    52.0  blue-collar      married  secondary             NaN          no   
9002    45.0   management      married   tertiary             NaN          no   
10211   51.0  blue-collar      married  secondary             NaN          no   
10363   33.0  blue-collar      married  secondary             NaN          no   
11066   28.0   technician       single  secondary             NaN          no   

      emprestimo   dia  mes  duracao  qtd_contato deposito  
936           no  13.0  nov     90.0          1.0      yes  
1213          no   9.0  feb    126.0          1.0      yes  
1735          no  30.0  apr    169.0          1.0      yes  
1871          no  11.0  may    733.0          2.0      yes  
3458          no  27.0  apr    370.0          1.0      yes  
5130          no   2.0  jun    384.0          1.0      yes  
5319          no  29.0  aug    488.0          2.0       no  
5857          no  13.0  may     29.0          4.0       no  
6080         yes  20.0  jun    109.0          3.0       no  
9002          no  30.0  jul    135.0          3.0       no  
10211         no   8.0  aug    237.0          4.0       no  
10363         no  20.0  nov     63.0          5.0       no  
11066         no   3.0  feb    226.0          2.0       no

In [113]:
df_exercicio.drop(df_exercicio.loc[df_exercicio['saldo_bancario'].isnull()].index,inplace = True)

In [114]:
df_exercicio.loc[df_exercicio['saldo_bancario'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [115]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           0.0000
estado_civil      0.0000
educacao          0.0000
saldo_bancario    0.0000
possui_casa       0.5113
emprestimo        1.7490
dia               1.6055
mes               1.5069
duracao           0.6368
qtd_contato       4.2784
deposito          0.0000
dtype: float64

## Limpeza de nulos: possui_casa

In [118]:
df_exercicio.loc[df_exercicio['possui_casa'].isnull()]

idade       emprego estado_civil   educacao  saldo_bancario possui_casa  \
20     28.0        admin.     divorced  secondary           785.0         NaN   
21     43.0    management       single   tertiary          2067.0         NaN   
22     43.0    management     divorced   tertiary           388.0         NaN   
23     43.0   blue-collar      married    primary          -192.0         NaN   
24     37.0    unemployed       single  secondary           381.0         NaN   
315    27.0        admin.      married  secondary           -97.0         NaN   
316    41.0    management       single   tertiary         11008.0         NaN   
317    42.0   blue-collar       single    primary          4930.0         NaN   
318    41.0  entrepreneur      married  secondary           707.0         NaN   
319    33.0    technician      married  secondary             4.0         NaN   
320    46.0        admin.      married  secondary           126.0         NaN   
321    41.0   blue-collar      married  secondary           671.0         NaN   
322    41.0   blue-collar      married    primary           219.0         NaN   
323    41.0   blue-collar     divorced  secondary          3622.0         NaN   
324    55.0       retired     divorced  secondary          1580.0         NaN   
325    41.0    technician      married  secondary          1319.0         NaN   
326    55.0    management      married    primary          2152.0         NaN   
327    39.0      services      married  secondary          2843.0         NaN   
328    40.0  entrepreneur      married   tertiary             0.0         NaN   
329    59.0       retired      married    primary          1115.0         NaN   
330    42.0   blue-collar       single  secondary           625.0         NaN   
331    41.0   blue-collar      married    primary           246.0         NaN   
332    27.0   blue-collar      married  secondary            61.0         NaN   
333    25.0   blue-collar       single  secondary          -191.0         NaN   
334    43.0    unemployed      married    primary           519.0         NaN   
335    44.0      services      married  secondary           522.0         NaN   
336    26.0   blue-collar       single  secondary           -46.0         NaN   
337    52.0       retired      married   tertiary          2269.0         NaN   
338    54.0      services      married  secondary           386.0         NaN   
339    56.0    management      married   tertiary             0.0         NaN   
340    57.0    management      married   tertiary          9103.0         NaN   
341    29.0   blue-collar       single  secondary           908.0         NaN   
342    47.0    technician      married   tertiary          1694.0         NaN   
343    57.0    management      married   tertiary           -92.0         NaN   
344    40.0   blue-collar      married  secondary           188.0         NaN   
345    46.0      services     divorced  secondary             0.0         NaN   
346    43.0   blue-collar       single  secondary           740.0         NaN   
347    45.0    management      married   tertiary           655.0         NaN   
348    24.0    management      married  secondary           867.0         NaN   
349    40.0   blue-collar      married  secondary          1009.0         NaN   
1095   41.0      services       single  secondary          1709.0         NaN   
1096   51.0        admin.       single   tertiary           394.0         NaN   
1097   47.0   blue-collar      married  secondary          1125.0         NaN   
1098   41.0   blue-collar      married    primary          -462.0         NaN   
1099   29.0    unemployed       single    primary           550.0         NaN   
1100   32.0    technician       single  secondary           540.0         NaN   
1101   26.0   blue-collar       single    primary           941.0         NaN   
1102   40.0    management       single   tertiary          1361.0         NaN   
1103   33.0      services       singl

In [120]:
df_exercicio.drop(df_exercicio.loc[df_exercicio['possui_casa'].isnull()].index,inplace=True)

In [121]:
df_exercicio.loc[df_exercicio['possui_casa'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [122]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           0.0000
estado_civil      0.0000
educacao          0.0000
saldo_bancario    0.0000
possui_casa       0.0000
emprestimo        1.7490
dia               1.6048
mes               1.5146
duracao           0.6401
qtd_contato       4.1922
deposito          0.0000
dtype: float64

## Limpeza de nulos: emprestimo

In [123]:
df_exercicio.loc[df_exercicio['emprestimo'].isnull()]

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
12     29.0   management      married   tertiary           199.0         yes   
13     46.0  blue-collar       single   tertiary           460.0         yes   
14     41.0   technician       single   tertiary           703.0         yes   
15     35.0   management     divorced   tertiary          3837.0         yes   
16     32.0  blue-collar       single    primary           611.0         yes   
...     ...          ...          ...        ...             ...         ...   
5695   42.0  blue-collar      married    primary           704.0         yes   
5696   24.0  blue-collar       single  secondary           476.0         yes   
5697   57.0      retired      married  secondary          3738.0         yes   
5698   35.0   management      married    primary           634.0          no   
5699   32.0   management       single   tertiary          1134.0          no   

     emprestimo   dia  mes  duracao  qtd_contato deposito  
12          NaN   NaN  may   1689.0          4.0      yes  
13          NaN   NaN  may   1102.0          2.0      yes  
14          NaN   NaN  may    943.0          2.0      yes  
15          NaN   NaN  may   1084.0          1.0      yes  
16          NaN   NaN  may    541.0          3.0      yes  
...         ...   ...  ...      ...          ...      ...  
5695        NaN  31.0  jul     88.0         10.0       no  
5696        NaN   4.0  jun    161.0          1.0       no  
5697        NaN  19.0  apr    220.0          1.0       no  
5698        NaN   7.0  jul    185.0          2.0       no  
5699        NaN  28.0  aug     85.0          4.0       no  

[194 rows x 12 columns]

In [124]:
df_exercicio.emprestimo.mode()

0    no
Name: emprestimo, dtype: object

In [126]:
df_exercicio.loc[df_exercicio['emprestimo'].isnull(),'emprestimo'] = df_exercicio.emprestimo.mode()[0]

In [127]:
df_exercicio.loc[df_exercicio['emprestimo'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [128]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           0.0000
estado_civil      0.0000
educacao          0.0000
saldo_bancario    0.0000
possui_casa       0.0000
emprestimo        0.0000
dia               1.6048
mes               1.5146
duracao           0.6401
qtd_contato       4.1922
deposito          0.0000
dtype: float64

## Limpeza de nulos: dia

In [129]:
df_exercicio.loc[df_exercicio['dia'].isnull()]

idade        emprego estado_civil   educacao  saldo_bancario  \
12     29.0     management      married   tertiary           199.0   
13     46.0    blue-collar       single   tertiary           460.0   
14     41.0     technician       single   tertiary           703.0   
15     35.0     management     divorced   tertiary          3837.0   
16     32.0    blue-collar       single    primary           611.0   
...     ...            ...          ...        ...             ...   
7570   32.0     management      married   tertiary           128.0   
7571   42.0     technician     divorced  secondary           305.0   
7572   46.0  self-employed      married   tertiary           -80.0   
7573   35.0     technician      married  secondary           198.0   
7574   41.0     management      married   tertiary          1170.0   

     possui_casa emprestimo  dia  mes  duracao  qtd_contato deposito  
12           yes         no  NaN  may   1689.0          4.0      yes  
13           yes         no  NaN  may   1102.0          2.0      yes  
14           yes         no  NaN  may    943.0          2.0      yes  
15           yes         no  NaN  may   1084.0          1.0      yes  
16           yes         no  NaN  may    541.0          3.0      yes  
...          ...        ...  ...  ...      ...          ...      ...  
7570         yes         no  NaN  NaN    320.0          3.0       no  
7571         yes         no  NaN  NaN     82.0          4.0       no  
7572         yes         no  NaN  NaN    112.0          2.0       no  
7573          no         no  NaN  NaN    282.0         13.0       no  
7574          no         no  NaN  NaN    200.0          4.0       no  

[178 rows x 12 columns]

In [130]:
df_exercicio.dia.mean()

15.693421293751145

In [131]:
round(df_exercicio.dia.mean())

16

In [132]:
df_exercicio.loc[df_exercicio['dia'].isnull(),'dia'] = round(df_exercicio.dia.mean())

In [133]:
df_exercicio.loc[df_exercicio['dia'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [134]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           0.0000
estado_civil      0.0000
educacao          0.0000
saldo_bancario    0.0000
possui_casa       0.0000
emprestimo        0.0000
dia               0.0000
mes               1.5146
duracao           0.6401
qtd_contato       4.1922
deposito          0.0000
dtype: float64

## Limpeza de nulos: mes

In [136]:
df_exercicio.loc[df_exercicio['mes'].isnull()]

idade      emprego estado_civil   educacao  saldo_bancario possui_casa  \
7491   42.0     services     divorced  secondary           466.0          no   
7492   58.0   technician      married  secondary          1463.0         yes   
7493   43.0   technician     divorced  secondary          6835.0          no   
7494   29.0   technician       single  secondary          -397.0         yes   
7495   39.0  blue-collar      married  secondary          5222.0         yes   
...     ...          ...          ...        ...             ...         ...   
9090   27.0     services       single  secondary          -151.0          no   
9091   25.0   management       single   tertiary          1602.0         yes   
9092   37.0       admin.      married  secondary          1314.0         yes   
9093   36.0   technician      married  secondary            84.0         yes   
9094   56.0      retired      married    primary          1906.0          no   

     emprestimo   dia  mes  duracao  qtd_contato deposito  
7491         no  16.0  NaN     48.0          1.0       no  
7492         no  16.0  NaN    110.0          3.0       no  
7493         no  16.0  NaN    187.0          1.0       no  
7494         no  16.0  NaN    114.0          1.0       no  
7495         no  16.0  NaN    102.0          1.0       no  
...         ...   ...  ...      ...          ...      ...  
9090         no   8.0  NaN    283.0          2.0       no  
9091         no  26.0  NaN    154.0          1.0       no  
9092         no  22.0  NaN     74.0          1.0       no  
9093         no  20.0  NaN     67.0          1.0       no  
9094        yes  19.0  NaN     45.0          9.0       no  

[168 rows x 12 columns]

In [137]:
df_exercicio.mes.mode()

0    may
Name: mes, dtype: object

In [139]:
df_exercicio.loc[df_exercicio['mes'].isnull(),'mes']=df_exercicio.mes.mode()[0]

In [140]:
df_exercicio.loc[df_exercicio['mes'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [141]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           0.0000
estado_civil      0.0000
educacao          0.0000
saldo_bancario    0.0000
possui_casa       0.0000
emprestimo        0.0000
dia               0.0000
mes               0.0000
duracao           0.6401
qtd_contato       4.1922
deposito          0.0000
dtype: float64

## Limpeza de nulos: duracao

In [144]:
df_exercicio.loc[df_exercicio['duracao'].isnull()]

idade        emprego estado_civil   educacao  saldo_bancario  \
2500   55.0     management      married    primary           113.0   
2501   33.0    blue-collar       single  secondary           498.0   
2502   43.0     technician      married   tertiary          2643.0   
2503   34.0       services      married  secondary          1337.0   
2504   35.0       services      married  secondary           618.0   
...     ...            ...          ...        ...             ...   
2566   38.0     technician      married  secondary         10596.0   
2567   22.0        student       single   tertiary           691.0   
2568   25.0   entrepreneur       single   tertiary           347.0   
2569   34.0  self-employed       single   tertiary          9962.0   
2570   27.0     technician       single   tertiary         11862.0   

     possui_casa emprestimo   dia  mes  duracao  qtd_contato deposito  
2500          no         no  12.0  aug      NaN          1.0      yes  
2501          no         no  12.0  aug      NaN          2.0      yes  
2502          no         no  12.0  aug      NaN          3.0      yes  
2503          no         no  12.0  aug      NaN          2.0      yes  
2504          no         no  12.0  aug      NaN          1.0      yes  
...          ...        ...   ...  ...      ...          ...      ...  
2566          no         no  19.0  aug      NaN          1.0      yes  
2567          no         no  19.0  aug      NaN          1.0      yes  
2568          no         no  20.0  aug      NaN          1.0      yes  
2569          no         no  20.0  aug      NaN          1.0      yes  
2570          no         no  20.0  aug      NaN          1.0      yes  

[71 rows x 12 columns]

In [145]:
df_exercicio.drop(df_exercicio.loc[df_exercicio['duracao'].isnull()].index,inplace=True)

In [146]:
df_exercicio.loc[df_exercicio['duracao'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [147]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0000
emprego           0.0000
estado_civil      0.0000
educacao          0.0000
saldo_bancario    0.0000
possui_casa       0.0000
emprestimo        0.0000
dia               0.0000
mes               0.0000
duracao           0.0000
qtd_contato       4.2192
deposito          0.0000
dtype: float64

## Limpeza de nulos: duracao

In [148]:
df_exercicio.loc[df_exercicio['qtd_contato'].isnull()]

idade       emprego estado_civil   educacao  saldo_bancario  \
1112    41.0    unemployed       single    primary           163.0   
1113    26.0      services       single  secondary           110.0   
1114    40.0    technician       single   tertiary          1646.0   
1115    45.0    unemployed      married  secondary          2040.0   
1116    55.0    technician      married  secondary          -308.0   
...      ...           ...          ...        ...             ...   
10195   29.0    technician       single  secondary            18.0   
10196   46.0        admin.       single  secondary           306.0   
10197   37.0  entrepreneur      married    primary             8.0   
10198   30.0    management      married   tertiary          -250.0   
10199   41.0   blue-collar      married  secondary           449.0   

      possui_casa emprestimo   dia  mes  duracao  qtd_contato deposito  
1112           no         no  16.0  jan    707.0          NaN      yes  
1113           no         no  16.0  jan   1136.0          NaN      yes  
1114          yes         no  16.0  feb    215.0          NaN      yes  
1115           no         no  16.0  feb    803.0          NaN      yes  
1116           no         no  16.0  feb    781.0          NaN      yes  
...           ...        ...   ...  ...      ...          ...      ...  
10195          no         no   3.0  feb    206.0          NaN       no  
10196         yes         no  11.0  may    289.0          NaN       no  
10197         yes        yes  16.0  jul     46.0          NaN       no  
10198         yes         no  20.0  jun     11.0          NaN       no  
10199         yes        yes  12.0  may    313.0          NaN       no  

[465 rows x 12 columns]

In [149]:
df_exercicio.qtd_contato.mean()

2.508241758241758

In [150]:
round(df_exercicio.qtd_contato.mean())

3

In [151]:
df_exercicio.loc[df_exercicio['qtd_contato'].isnull(),'qtd_contato']=round(df_exercicio.qtd_contato.mean())

In [152]:
df_exercicio.loc[df_exercicio['qtd_contato'].isnull()]

Empty DataFrame
Columns: [idade, emprego, estado_civil, educacao, saldo_bancario, possui_casa, emprestimo, dia, mes, duracao, qtd_contato, deposito]
Index: []

In [153]:
round(df_exercicio.isna().sum() / len(df_exercicio) * 100, 4)

idade             0.0
emprego           0.0
estado_civil      0.0
educacao          0.0
saldo_bancario    0.0
possui_casa       0.0
emprestimo        0.0
dia               0.0
mes               0.0
duracao           0.0
qtd_contato       0.0
deposito          0.0
dtype: float64